### ISSUE TITLE
Project unload throws NUnitEngineException

### ISSUE THREAD

**Issue by JensLeible** [TestCentric/testcentric-gui#720]
I get this exception everytime the project unloads or reloads (unload issue there, too):

> See the end of this message for details on invoking 
> just-in-time (JIT) debugging instead of this dialog box.
> 
> ************** Exception Text **************
NUnit.Engine.NUnitEngineUnloadException: Exception encountered unloading application domain ---> NUnit.Engine.NUnitEngineException: Exception encountered unloading application domain: Error while unloading appdomain. (Exception from HRESULT: > 0x80131015)
> Application domain name: domain-
> Application domain BaseDirectory: D:\_Projects\MIT_EAF\SQA\UserStories\US171383_LLM_OverrideStateRetention\Implementation\bin\
> 
>    --- End of inner exception stack trace ---
> 
> Server stack trace: 
>    at TestCentric.Engine.Services.DomainManager.DomainUnloader.Unload() in C:\Users\charlie\dev\TestCentric\testcentric-gui\src\TestEngine\testcentric.engine.core\Services\DomainManager.cs:line 161
>    at TestCentric.Engine.Runners.TestDomainRunner.UnloadPackage() in C:\Users\charlie\dev\TestCentric\testcentric-gui\src\TestEngine\testcentric.engine.core\Runners\TestDomainRunner.cs:line 39
>    at TestCentric.Engine.Runners.AbstractTestRunner.Unload() in C:\Users\charlie\dev\TestCentric\testcentric-gui\src\TestEngine\testcentric.engine.core\Runners\AbstractTestRunner.cs:line 152
   at TestCentric.Engine.Communication.Transports.Remoting.TestAgentRemotingTransport.Unload() in > C:\Users\charlie\dev\TestCentric\testcentric-gui\src\TestEngine\testcentric.engine.core\Communication\Transports\Remoting\TestAgentRemotingTransport.cs:line 130
>    at System.Runtime.Remoting.Messaging.StackBuilderSink._PrivateProcessMessage(IntPtr md, Object[] args, Object server, Object[]& outArgs)
>    at System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage(IMessage msg)
> 
> Exception rethrown at [0]: 
>    at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
>    at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
>    at TestCentric.Engine.ITestEngineRunner.Unload()
>    at TestCentric.Engine.Runners.ProcessRunner.UnloadPackage() in C:\Users\charlie\dev\TestCentric\testcentric-gui\src\TestEngine\testcentric.engine\Runners\ProcessRunner.cs:line 104
>    at TestCentric.Engine.Runners.AbstractTestRunner.Unload() in C:\Users\charlie\dev\TestCentric\testcentric-gui\src\TestEngine\testcentric.engine.core\Runners\AbstractTestRunner.cs:line 153
>    at TestCentric.Engine.Runners.MasterTestRunner.UnloadPackage() in C:\Users\charlie\dev\TestCentric\testcentric-gui\src\TestEngine\testcentric.engine\Runners\MasterTestRunner.cs:line 341
>    at TestCentric.Gui.Model.TestModel.ReloadTests()
>    at TestCentric.Gui.Presenters.TestCentricPresenter.<WireUpEvents>b__9_23()
>    at System.Windows.Forms.ToolStripItem.RaiseEvent(Object key, EventArgs e)
>    at System.Windows.Forms.ToolStripMenuItem.OnClick(EventArgs e)
>    at System.Windows.Forms.ToolStripItem.HandleClick(EventArgs e)
>    at System.Windows.Forms.ToolStripItem.HandleMouseUp(MouseEventArgs e)
>    at System.Windows.Forms.ToolStrip.OnMouseUp(MouseEventArgs mea)
>    at System.Windows.Forms.ToolStripDropDown.OnMouseUp(MouseEventArgs mea)
>    at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
>    at System.Windows.Forms.Control.WndProc(Message& m)
>    at System.Windows.Forms.ToolStrip.WndProc(Message& m)
>    at System.Windows.Forms.ToolStripDropDown.WndProc(Message& m)
>    at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
> 
> 
> ************** Loaded Assemblies **************
> mscorlib
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.8.4300.0 built by: NET48REL1LAST_C
>     CodeBase: file:///C:/Windows/Microsoft.NET/Framework64/v4.0.30319/mscorlib.dll
> ----------------------------------------
> testcentric
>     Assembly Version: 1.6.2.0
>     Win32 Version: 1.6.2
>     CodeBase: file:///C:/ProgramData/chocolatey/lib/testcentric-gui/tools/testcentric.exe
> ----------------------------------------
> TestCentric.Gui.Runner
>     Assembly Version: 1.6.2.0
>     Win32 Version: 1.6.2
>     CodeBase: file:///C:/ProgramData/chocolatey/lib/testcentric-gui/tools/TestCentric.Gui.Runner.DLL
> ----------------------------------------
> TestCentric.Common
>     Assembly Version: 1.6.2.0
>     Win32 Version: 1.6.2
>     CodeBase: file:///C:/ProgramData/chocolatey/lib/testcentric-gui/tools/TestCentric.Common.DLL
> ----------------------------------------
> TestCentric.Gui.Model
>     Assembly Version: 1.6.2.0
>     Win32 Version: 1.6.2
>     CodeBase: file:///C:/ProgramData/chocolatey/lib/testcentric-gui/tools/TestCentric.Gui.Model.DLL
> ----------------------------------------
> System.Windows.Forms
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.8.4270.0 built by: NET48REL1LAST_C
>     CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms/v4.0_4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll
> ----------------------------------------
> System
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.8.4300.0 built by: NET48REL1LAST_C
>     CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System/v4.0_4.0.0.0__b77a5c561934e089/System.dll
> ----------------------------------------
> System.Drawing
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.8.3761.0 built by: NET48REL1
>     CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Drawing/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll
> ----------------------------------------
> nunit.engine.api
>     Assembly Version: 3.0.0.0
>     Win32 Version: 3.11.1
>     CodeBase: file:///C:/ProgramData/chocolatey/lib/testcentric-gui/tools/nunit.engine.api.DLL
> ----------------------------------------
> testcentric.engine
>     Assembly Version: 1.6.2.0
>     Win32 Version: 1.6.2
>     CodeBase: file:///C:/ProgramData/chocolatey/lib/testcentric-gui/tools/testcentric.engine.DLL
> ----------------------------------------
> testcentric.engine.core
>     Assembly Version: 1.6.2.0
>     Win32 Version: 1.6.2
>     CodeBase: file:///C:/ProgramData/chocolatey/lib/testcentric-gui/tools/testcentric.engine.core.DLL
> ----------------------------------------
> System.Xml
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.8.3761.0 built by: NET48REL1
>     CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
> ----------------------------------------
> System.Configuration
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.8.4190.0 built by: NET48REL1LAST_B
>     CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Configuration/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll
> ----------------------------------------
> System.Core
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.8.4320.0 built by: NET48REL1LAST_C
>     CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Core/v4.0_4.0.0.0__b77a5c561934e089/System.Core.dll
> ----------------------------------------
> System.Web
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.8.4330.0 built by: NET48REL1LAST_B
>     CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_64/System.Web/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Web.dll
> ----------------------------------------
> System.Runtime.Remoting
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.8.3761.0 built by: NET48REL1
>     CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Runtime.Remoting/v4.0_4.0.0.0__b77a5c561934e089/System.Runtime.Remoting.dll
> ----------------------------------------
> TestCentric.Gui.Components
>     Assembly Version: 0.0.0.0
>     Win32 Version: 0.0.0.0
>     CodeBase: file:///C:/ProgramData/chocolatey/lib/testcentric-gui/tools/TestCentric.Gui.Components.DLL
> ----------------------------------------
> nunit.uiexception
>     Assembly Version: 1.6.2.0
>     Win32 Version: 1.6.2
>     CodeBase: file:///C:/ProgramData/chocolatey/lib/testcentric-gui/tools/nunit.uiexception.DLL
> ----------------------------------------
> testcentric.engine.metadata
>     Assembly Version: 1.6.2.0
>     Win32 Version: 1.6.2
>     CodeBase: file:///C:/ProgramData/chocolatey/lib/testcentric-gui/tools/testcentric.engine.metadata.DLL
> ----------------------------------------
> Microsoft.GeneratedCode
>     Assembly Version: 1.0.0.0
>     Win32 Version: 4.8.3761.0 built by: NET48REL1
>     CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
> ----------------------------------------
> 
> ************** JIT Debugging **************
> To enable just-in-time (JIT) debugging, the .config file for this
> application or computer (machine.config) must have the
> jitDebugging value set in the system.windows.forms section.
> The application must also be compiled with debugging
> enabled.
> 
> For example:
> 
> <configuration>
>   <system.windows.forms jitDebugging="true" />
> </configuration>
> 
> When JIT debugging is enabled, any unhandled exception
> will be sent to the JIT debugger registered on the computer
> rather than be handled by this dialog box.

**To Reproduce**
1. Load NUnit 3 test project DLL
2. Click on File>Close
3. See error

**Expected behavior**
No exception during unload.

**Environment:**
 - TestCentric Version 1.6.2
 - OS: Windows 10
 - Target Runtime for your test assembly: 4.6.1
 - Installed via chocolatey

**Comment by JensLeible** [comment]
**Environment (please complete the following information):**
 - TestCentric Version 1.6.2
 - OS: Windows 10
 - Target Runtime for your test assembly: 4.6.1

**Comment by CharliePoole** [comment]
I'm unable to reproduce this error following the instructions included. Are there any other steps needed?

**Comment by JensLeible** [comment]
Unfortunately not. How can I support you?

**Comment by CharliePoole** [comment]
Thinking about it, I probably couldn't reproduce it because my test assemblies all unloaded properly. It needs an assembly, which refuses to be unloaded. The GUI should ignore that error, possibly giving a warning. Let me see what I can come up with.

**Comment by CharliePoole** [comment]
I have confirmed that this is a problem when the test itself cannot be unloaded.

When an `AppDomain` can't be unloaded an `NUnitEngineUnloadException` is thrown. This is intended to be handled by the runner, which should normally let the user know there was a problem but not cancel the run or terminate the application.

The GUI handles this exception correctly if it takes place when the application is closed but not when a test assembly is deliberately unloaded or reloaded from the menu.

I have categorized this as High Priority, which means it will be fixed for the next release, currently planned to be 2.0.
